### ISSUE TITLE
How to Visualization the data

### ISSUE THREAD

**Issue by liuwei960620** [pmeier/pystiche#481]
How to Visualization the data about content_loss and style_loss?    Is there a special way with pystiche or same as common pytorch?

**Comment by pmeier** [comment]
Hey @liuwei960620. I'm sorry, but I don't understand your question. What do you mean with

> data

here? The images? If so, you can use `pystiche.image.show_image` to display them. 

If you want to know what your `content_loss` and `style_loss` comprise you can just print them:

```python
from pystiche import enc, ops

mle = enc.vgg19_multi_layer_encoder()

style_layers = ("relu1_1", "relu2_1", "relu3_1", "relu4_1", "relu5_1")
style_weight = 1e3


def get_style_op(encoder, layer_weight):
    return ops.GramOperator(encoder, score_weight=layer_weight)


style_loss = ops.MultiLayerEncodingOperator(
    mle, style_layers, get_style_op, score_weight=style_weight,
)
print(style_loss)
```

```
MultiLayerEncodingOperator(
  encoder=VGGMultiLayerEncoder(arch=vgg19, framework=torch, allow_inplace=True),
  score_weight=1000
  (relu1_1): GramOperator(score_weight=0.2)
  (relu2_1): GramOperator(score_weight=0.2)
  (relu3_1): GramOperator(score_weight=0.2)
  (relu4_1): GramOperator(score_weight=0.2)
  (relu5_1): GramOperator(score_weight=0.2)
)
```

**Comment by liuwei960620** [comment]
Hey Philip Meier,Thanks for your reply。It's difficult to insert a figure on github ,so I express in e-mail.
I'am a beginner.I want to output these numbers for example as figure 1
  I   can use plot( ) with common pytorch.But with pystiche,I don't know how to do?
Thank you .



刘伟
Mobile:139-5218-9127
Email:liuwei.xz@163.com
 
From: Philip Meier
Date: 2021-01-07 17:34
To: pmeier/pystiche
CC: liuwei960620; Mention
Subject: Re: [pmeier/pystiche] How to Visualization the data (#481)
Hey @liuwei960620. I'm sorry, but I don't understand your question. What do you mean with
data
here? The images? If so, you can use pystiche.image.show_image to display them.
If you want to know what your content_loss and style_loss comprise you can just print them:
from pystiche import enc, ops
mle = enc.vgg19_multi_layer_encoder()
style_layers = ("relu1_1", "relu2_1", "relu3_1", "relu4_1", "relu5_1")
style_weight = 1e3
def get_style_op(encoder, layer_weight):
    return ops.GramOperator(encoder, score_weight=layer_weight)
style_loss = ops.MultiLayerEncodingOperator(
    mle, style_layers, get_style_op, score_weight=style_weight,
)
print(style_loss)
MultiLayerEncodingOperator(
  encoder=VGGMultiLayerEncoder(arch=vgg19, framework=torch, allow_inplace=True),
  score_weight=1000
  (relu1_1): GramOperator(score_weight=0.2)
  (relu2_1): GramOperator(score_weight=0.2)
  (relu3_1): GramOperator(score_weight=0.2)
  (relu4_1): GramOperator(score_weight=0.2)
  (relu5_1): GramOperator(score_weight=0.2)
)

―

**Comment by liuwei960620** [comment]
![微信图片_202101081558061](https://user-images.githubusercontent.com/77082988/103991404-8d8e7080-51cd-11eb-88d6-609d8b58a8c8.png)

**Comment by pmeier** [comment]
Ah ok, what you want is to plot the content and style loss over the number of iterations. For that we need to write a small optimization loop ourselves:

```python
optimizer = optim.default_image_optimizer(input_image)

num_steps = 500
losses = []

for _ in range(num_steps):

    def closure():
        optimizer.zero_grad()

        loss = criterion(input_image)
        loss.backward()

        losses.append(loss)
        return float(loss)

    optimizer.step(closure)
```

Here I assumed you already have an `input_image` and `criterion` (perceptual loss) set up. You can follow [this usage example](https://pystiche.readthedocs.io/en/stable/galleries/examples/beginner/example_nst_with_pystiche.html#sphx-glr-galleries-examples-beginner-example-nst-with-pystiche-py) up to the call to `optim.image_optimization` if you don't know how to do this.

---

The `criterion` does not return a tensor, but rather a fine-grained dictionary of losses:

```python
print(losses[0])
```

```
content_loss      : 0.000e+00
style_loss.relu1_1: 6.081e+01
style_loss.relu2_1: 4.697e+02
style_loss.relu3_1: 2.603e+03
style_loss.relu4_1: 1.190e+03
style_loss.relu5_1: 4.465e+01
```

Since you don't want to know style loss on each layer individually, we can use the `aggregate` method:

```python
print(losses[0].aggregate(1))
```

```
content_loss: 0.000e+00
style_loss  : 4.369e+03
```

Afterward we can extract the content and style losses recorded during the optimization:

```python
losses = [loss.aggregate(1) for loss in losses]
content_losses, style_losses = zip(
    *[(loss["content_loss"], loss["style_loss"]) for loss in losses]
)
```

---

As a last step we can use any plotting package to plot them for us. I'm using [`matplotlib`]() for this:

```python
import matplotlib.pyplot as plt

plt.semilogy(content_losses)
plt.semilogy(style_losses)
plt.legend(["content_loss", "style_loss"])
plt.grid()
plt.show()
```

![content_style_loss](https://user-images.githubusercontent.com/6849766/104028070-90d32d80-51c8-11eb-9757-e01951384055.png)

---

Let me know if this answers your question. In retrospect, I probably should write up an example for this use case.

**Comment by liuwei960620** [comment]
@pmeier ,Ok,It's just what I need.I can easily compare it to other methods.Thank you very much!
